src/consensus.rs

```diff
@@ -486,149 +477,154 @@ impl Consensus {
             }
         }
 
+        // If this is the origin peer of the cluster, try to add origin peer to consensus
+        if let Err(err) = self.try_add_origin() {
+            log::error!("Failed to add origin peer to consensus: {err}");
+        }
+
         let tick_period = Duration::from_millis(self.config.tick_period_ms);
         let mut previous_tick = Instant::now();
+        let mut idle_cycles = 0_usize;
 
         loop {
-            // Apply in-memory changes to the Raft State Machine
-            // If updates = None, we need to skip this step due to timing limits
-            // If updates = Some(0), means we didn't receive any updates explicitly
-            let updates = self.advance_node(previous_tick, tick_period)?;
+            // Wait (for up to `tick_period`) for incoming client requests and Raft messages
+            let raft_messages = self.advance_node(tick_period)?;
 
-            let mut elapsed = previous_tick.elapsed();
+            // Calculate how many ticks passed since the last one
+            let elapsed_ticks = previous_tick.elapsed().div_duration_f32(tick_period) as u32;
 
-            while elapsed > tick_period {
-                self.node.tick();
+            // Update previous tick timestamp
+            previous_tick += tick_period * elapsed_ticks;
 
-                previous_tick += tick_period;
-                elapsed -= tick_period;
-            }
+            // Calculate how many ticks we should *report* to Raft node.
+            //
+            // If last iteration of the loop took too long to complete, and we report all elapsed
+            // ticks to Raft node, it might trigger unnecessary leader election.
+            //
+            // To prevent this, we check if we received new Raft messages (i.e., we are still
+            // connected to Raft leader), and cap how many ticks we report to Raft node.
+            //
+            // By default, election is triggered if no Raft messages were received for 20 ticks,
+            // so we report at most 15 ticks.
+            //
+            // See https://docs.rs/raft/latest/raft/struct.Config.html#structfield.election_tick.
+            let report_ticks = if raft_messages > 0 {
+                // Default `election_tick` is 20, so expected value here is 15
+                let max_elapsed_ticks =
+                    cmp::max(1, self.raft_config.election_tick.saturating_sub(5));
 
-            if self.node.has_ready() {
-                // Persist AND apply changes, which were committed in the Raft State Machine
-                let stop_consensus = self.on_ready()?;
+                cmp::min(elapsed_ticks, max_elapsed_ticks as u32)
+            } else {
+                elapsed_ticks
+            };
```

Comment:
Should we have an upper bound? 

I mean, what if this node has been stuck (somehow) for a long while and we still report just 15 ticks? In that case I do expect new election.

Or am I missing an important detail here?

Reply:
The election will be triggered on the next few iterations of consensus loop then.